Add an extraction of function implementations

ecklbarb opened the pull request:
I add a new extraction class to extract the function implementations, because @nischikata needs the data to compute the semantic coupling between functions or files.

ecklbarb committed:
Add an extraction of function implementations

Add extraction of function implementations from the codeface database
into a "implementations.list" file. Before writing the .list-file, we
remove all linebreaks, tabs, and unknown symbols from the source code.

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

clhunsen commented on codeface_extraction/extractions.py:
```diff
@@ -266,6 +266,47 @@ def __init__(self, dbm, conf, resdir, csv_writer):
+        # for subclasses
+        self.sql = """
+                    SELECT c.id,
+                           c.commitHash, c.ChangedFiles,
```
Do you really need `c.ChangedFiles` here? This is no blocker, I am just wondering. :wink:

clhunsen commented on codeface_extraction/extractions.py:
```diff
@@ -266,6 +266,47 @@ def __init__(self, dbm, conf, resdir, csv_writer):
+                    AND p.analysisMethod = '{tagging}'
+                    AND cd.file IS NOT NULL
+
+                    ORDER BY cd.file, c.id, c.commitHash, cd.entityId
```
Wouldn't it be better for debugging to have another sorting here? E.g., `ORDER BY c.id, c.commitHash, cd.file, cd.entityId` or `ORDER BY cd.file, cd.entityId, c.id, c.commitHash`. In the end, my suggestion is to sort by (not in the SQL  sense!) file *and* function name.

clhunsen reviewed (approved):
This PR looks good. Thanks for the addition, @ecklbarb!

I added two comments on minor stuff. Maybe, you can respond to those before I merge.

bockthom commented:
@ecklbarb Thanks for your helpful changes!

In addition to Claus' comments, I have also another question: Currently, you extract those imlementation list files only on project level, but not on range level (as we do for all the other extractions).

Wouldn't it be helpful to have this also on range level? That might be useful for debugging and will be more consistent with the rest of the extractions. In addition, we would need that for release-based analyses, too. (Currently, we do not process any release-related information, but I still plan to implement some release-related analyses somewhen...)
So, it would be very easy to add the newly introduced extraction also on range-level.
@clhunsen What do you think about that?

bockthom commented:
FYI: I currently implement the extraction of commit messages (together with their commit hashes) into a separate file. I will open an additional pull request for the extraction after this PR is merged. Just that you are prepared for that :wink:

bockthom commented:
@ecklbarb I just remembered: Don't forget to update the copyright header. Thanks!

clhunsen commented:
> So, it would be very easy to add the newly introduced extraction also on range-level.
> @clhunsen What do you think about that?

This is a good idea. I thought about that too, but got distracted in that moment and forgot about it then. :laughing: 

> @ecklbarb I just remembered: Don't forget to update the copyright header. Thanks!

Haha, I *always* forget that one! :smile::man_facepalming: Nice catch, @bockthom.

nischikata commented on codeface_extraction/extractions.py:
```diff
@@ -266,6 +266,47 @@ def __init__(self, dbm, conf, resdir, csv_writer):
+            newImpl = newImpl.replace("\r", " ")
+            newImpl = newImpl.replace("\t", " ")
+            newImpl = newImpl.replace(u"\ufffd", " ")
+
```
instead of lines 301 - 304 I sugguest the following (NOTE: requires `import unicodedata, re`):
```python
"""  Removes control characters such as \r\n \x1b \ufffd from string impl and returns a unicode string where all control characters have been replaced by a space """

    newImpl = re.sub(r'\\ufff.', ' ', impl)
    newImpl = "".join(ch if unicodedata.category(ch)[0]!="C" else " " for ch in newImpl.decode('unicode-escape'))
```

clhunsen replied:
This looks great. Thanks, @nischikata! Didn't know that there is a good approach for doing it in a more generic way. I just would like to watch for consistent quotation marks and spaces around operators.

I propose to give also more explanatory comments, such as giving the following links:
- https://www.fileformat.info/info/unicode/category/index.htm and
- https://www.compart.com/en/unicode/block/U+FFF0.

clhunsen replied:
I tested this and it is unfortunately not working as expected: The output gets scrambled and control characters are not removed:

Input: `öffi �￼˜`
Output: `Ã¶ffi ï¿½ï¿¼Ë`
Expected output: <code>öffi&nbsp;&nbsp;&nbsp;&nbsp;</code>

I talked to @nischikata and we agree that there is something going wrong already when writing the code into the database. We also agree that we should be input-tolerant and support the following both scenarios: 
1. The database contents are correctly encoded and we should be able to remove problematic characters.
2. The database contents are scrambled already and we should still be able to remove problematic characters.

We should talk about this sometime soon.

ecklbarb commented:
The "implementations.list" files of small projects can become already large. For example, the file of Zeppelin is 100 MB and  of busybox is 6 MB.
I think, we shouldn't save the data twice.

@clhunsen Do you know, how I can add a flag to the extraction command, because it should be optional  to extract the implementation?

ecklbarb committed:
Minor changes in the extraction of the implementation

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>
Signed-off-by: Tina Schuh <schuht@fim.uni-passau.de>

ecklbarb committed:
Add new CLI argument for switching on the implementation extraction

Add the new boolean CLI argument "-i" or "--implementation" to switch on
or off the extraction of function implementations. The default value
is "False" (no implementation extraction).

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

clhunsen commented on codeface_extraction/codeface_extraction.py:
```diff
@@ -92,6 +92,8 @@ def get_parser():
     run_parser.add_argument('-p', '--project', help="Project configuration file",
                             required=True)
+    run_parser.add_argument('-i', '--implementation', help="Enable extraction of the source code of functions",
+                            type=lambda x: (str(x).lower() == 'true'), default=False)
```
Please use `action='store_true'` (see [here](https://docs.python.org/2.7/library/argparse.html#action) for more details).

clhunsen commented on codeface_extraction/codeface_extraction.py:
```diff
@@ -107,11 +109,12 @@ def run():
     # - First make all the args absolute
     __resdir = abspath(args.resdir)
     __codeface_conf, __project_conf = map(abspath, (args.config, args.project))
+    __exclude_impl = args.implementation
```
`!args.implementation`, I guess. If the `--implementation` parameter is given, you do **not** want to _exclude_ the function implementations. Otherwise, the name of the variable would be misleading.

clhunsen commented on codeface_extraction/extractions.py:
```diff
@@ -30,19 +33,22 @@
     __extractions = []
     for cls in Extraction.__subclasses__():
-        __extractions.append(cls(dbm, conf, resdir, csv_writer))
+        if (exclude_impl or
```
`!exclude_impl`, isn't it? This is likely a follow-up thing from the potential mistake with the new parser argument.

clhunsen commented on codeface_extraction/extractions.py:
```diff
@@ -292,18 +297,22 @@ def __init__(self, dbm, conf, resdir, csv_writer):
+        for (commitId, commitHash, fileId, entityId, impl) in result:
+            newImpl = impl.encode('utf-8')
+            newImpl = re.sub(r'\\ufff.', ' ', newImpl)
+            newImpl = "".join(ch if unicodedata.category(ch)[0]!="C" else " " for ch in newImpl.decode('unicode-escape'))
```
Please make quotation marks and spaces around operators more consistent with other parts of the code and insert some small documentation as suggested in [my earlier comment](https://github.com/se-passau/codeface-extraction/pull/14/files/efea611320c7382f0718ab2ec0371a7ceb5b9158..4c92249431392c6a0bde3b9b1e6b902147c31855#r201637899).

ecklbarb replied:
@nischikata, can you help me to write a small documentation?

clhunsen commented on codeface_extraction/extractions.py:
```diff
@@ -14,13 +14,16 @@
 #
 # Copyright 2015-2018 by Claus Hunsen <hunsen@fim.uni-passau.de>
 # Copyright 2016 by Thomas Bock <bockthom@fim.uni-passau.de>
+# Copyright 2018 by Barbara Eckl <ecklbarb@fim.uni-passau.de>
```
The copyright header in the other file is still missing.

clhunsen reviewed (changes requested):
Unfortunately, the new CLI parameter may not work as expected. Please check this thoroughly. 

Additionally, I added comments on some minor stuff.

ecklbarb committed:
Add new parameter for range level and add FunctionImplementationRangeExtraction

Add a new CLI parameter (-r or --range) to enable the extraction of
mails, commits and implementation on range level.

Add the extraction of function implementations on range level and minor
changes  of spaces, quotation marks and documentation.

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

clhunsen committed:
Refactor removal of special characters in implementation extractions

For better code re-use, the removal of special characters is now
extracted to the function 'remove_problematic_characters'.

Signed-off-by: Claus Hunsen <hunsen@fim.uni-passau.de>

clhunsen commented on codeface_extraction/extractions.py:
```diff
@@ -492,3 +498,57 @@ def __init__(self, dbm, conf, resdir, csv_writer):
+
+            newResult.append((commitId, commitHash, fileId, entityId, newImpl))
+
+        return newResult
```
Maybe, we can do something against the code duplication and extract the relevant parts into a separate global function. When trying out the routine, I hacked a patch which we could apply: https://gist.github.com/clhunsen/e0605b9910e81d88c42edd3e8ae3dc71

nischikata commented on codeface_extraction/extractions.py:
```diff
@@ -442,3 +485,69 @@ def __init__(self, dbm, conf, resdir, csv_writer):
+# HELPER FUNCTIONS
+#
+
+def remove_problematic_characters(text):
```
Here's my re-write of this function, which reconstructs the original (pre-database) string before removing problematic characters. It works with Claus' examples and random affected busybox samples, but needs further testing -- especially for source files that use non-utf8 encoding.
```python
def remove_problematic_characters(text):
    if isinstance(text, unicode):
        # deal with encoding (NOTE: Back-up plan instead of next 3 lines use the FTFY library to convert to most likely encoding)
        new_text = text.encode("unicode-escape")
        new_text, x = codecs.escape_decode(new_text, "utf-8")
        new_text = unicode(new_text, "utf-8", "replace")

        # remove unicode replacement characters
        new_text = re.sub(r"\\ufff.", " ", new_text.encode("unicode-escape"))

    else: # TODO: does this ever apply? and if yes, does it (still) work?

        new_text = text.encode("utf-8")
        new_text = re.sub(r"\\ufff.", " ", new_text.decode("utf-8")).encode("utf-8")

    # remove control characters and emojis
    new_text = u"".join(ch if unicodedata.category(ch)[0] != "C" else " " for ch in new_text.decode("unicode-escape"))

    return new_text
```

nischikata replied:
newest insight: the re-write works for files without encoding, but not for files have other encoding than utf-8 (we tested WIN1252 and latin-1)

nischikata committed:
Refactor function "remove_problematic_characters" to deal better with encoding

Fixed function in order to correctly handle the encoding applied by
codeface prior to writing the contents to the data base. The fix is
specifically for content that is originally encoded in utf-8 and is
tailored to the unicode conversion in codeface (function
"encode_as_utf8").

NOTE: The fix might work as expected for content that stems from other
than utf-8 encoded source files. Known errors with non-utf-8 encoded
source files: Any umlauts will be removed.

Signed-off-by: Tina Schuh <schuht@fim.uni-passau.de>

nischikata committed:
Reformat code

Minor source code formatting by PyCharm to meet the requirements of code
style.

Signed-off-by: Tina Schuh <schuht@fim.uni-passau.de>

nischikata committed:
Refactor function "remove_problematic_characters" to handle all encoding types

Added external library 'ftfy' ('fixes text for you') to detect and
assign correct encoding. Removed obsolete code parts, including the
check if input is a unicode string.
Successfully tested for files that contain umlauts, in the following
encodings: UTF-8, WINDOWS-1252, ISO-8859-1, no encoding.

IMPORTANT NOTE: Requires installation of Python 2.7 compatible version
of ftfy package via
pip install 'ftfy<5'

Signed-off-by: Tina Schuh <schuht@fim.uni-passau.de>

bockthom commented on codeface_extraction/extractions.py:
```diff
@@ -14,13 +14,18 @@
 # Copyright 2015-2018 by Claus Hunsen <hunsen@fim.uni-passau.de>
 # Copyright 2016 by Thomas Bock <bockthom@fim.uni-passau.de>
+# Copyright 2018 by Barbara Eckl <ecklbarb@fim.uni-passau.de>
+# Copyright 2018 by Tina Schuh <schuhtb@fim.uni-passau.de>
```
I just recognized that the "b" in Tina's e-mail address is wrong. Should be removed, I guess.

ecklbarb committed:
Correct e-mail-address in copyright header

Signed-off-by: Barbara Eckl <ecklbarb@fim.uni-passau.de>

clhunsen commented:
@bockthom, if you do not have reasons against merging anymore, you are free to merge.

bockthom commented:
> bockthom, if you do not have reasons against merging anymore, you are free to merge.

I just had a quick look at it, looks good. As I assume that all of you have tested the changes of this PR extensively, I will merge now.
